lib/flake-info.nix

```diff
@@ -16,5 +16,8 @@ finalLib: prevLib: # lib overlay
     versionSuffix =
       ".${finalLib.substring 0 8 (self.lastModifiedDate or "19700101")}.${self.shortRev or "dirty"}";
     revisionWithDefault = default: self.rev or default;
+
+    isFlake = true;
+    nixpkgsStorePathString = toString self;
```

Comment:
Needs a comment why this is needed and correct.

Reply:
No, the reasoning for why this is here is written down in the comment itself. Without this attribute, we cannot ship https://github.com/NixOS/nixpkgs/pull/254405. 

That is why we are doing this. Maybe it should be somewhere internal, I'm indifferent about that. But we do need it for good UX in NixOS flake systems. This particular use case does involve copying the whole nixpkgs into the store, unavoidably, since it will wind up in a derivation; it's not intended to be used for imports as we can already do that with pkgs.path, modulesPath, etc etc. 

I'm personally lightly pessimistic about lazy trees getting in, but regardless of my own views on this, it would be annoying to gate the UX improvement that many people want on a new Nix version that will take some time to stabilize and shake out the regressions. 

I do acknowledge this is extremely cursed code that depends on a load of C++nix implementation details and I would like to minimize the amount that it will break in the foreseeable future.

Reply:
We need some way to inject the dependency to NixOS so we can put the nixpkgs into the resulting closure so the UX of nix-shell and nix shell isn't awful. My last attempt was just injecting it directly into NixOS, but I saw this infrastructure that felt more appropriate.

But now I wonder if injecting it directly into NixOS is the least bad option; the disadvantage of that one is this infrastructure not being reusable, but perhaps that's for the better since it's rather cursed?

Reply:
> depends on a load of C++nix implementation details

I don't think path value semantics can be considered implementation details.

Path values have been an integral part of the language since basically forever. Any Nix implementation needs to implement path value semantics. I don't think the lazy-trees author appreciated that when they wrote the branch.
The problems with `self.outPath` and in-flake `./.` are squarely with Flakes.

> gate the UX improvement that many people want on a new Nix version

This is unfortunate, but I don't think it's an excuse for making a mess in `nixpkgs`.

> wonder if injecting it directly into NixOS is the least bad option

I'll take a messy implementation over a messy interface any day.

Reply:
> ship #254405.

For context, this can't be done correctly and I don't think we should try to pretend that we have a mechanism that forwards `pkgs` to other evaluations on the machine for which it was used. Usually I support incremental improvements, but if they can't lead to a good end state, the approach should be reconsidered in its entirety.

Reply:
Hmmm wait a minute. Wait a minute. We can probably actually fix this properly by forwarding the metadata of the flake input locking into the registry entry. 🤦🏻‍♀️ which would actually do this properly *and* not add to closure size, since we can forward to the flake registry for NIX_PATH. 

The only thing we lose there is *sending* nixpkgs to the new installation automatically so it works offline, rather than just merely putting the metadata in to get the right version lazily.

Reply:
Hmm. Nope. You can't get info of your own lock file entry from inside a flake. This is .... probably reasonable, because it would enable some truly spectacular spooky action at a distance, but it is quite sad for this purpose.

In conclusion, we still need something like this.